File: Day 4/Day 4.py
def parse_input(filename: str):
    with open(filename, "r") as file:
        lines = file.read()
        list_of_dictionaries = []
        # Split 1: Separates different entries
        lines = lines.strip().split("\n\n")
        for line in lines:
            # Split 2: Separates different fields within a single entry
            line = line.split()
            dictionary = {}
            for item in line:
                # Split 3: Separates fields from their values
                item = item.split(":")
                dictionary[item[0]] = item[1]
            list_of_dictionaries.append(dictionary)
        return list_of_dictionaries


def validate_entry_count(entry: dict):
    required_keys = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
    print("Checking entry: ", entry)
    for key in required_keys:
        if key not in entry:
            print("This entry is missing key: ", key)
            print()
            return False
    print("No (important) missing keys detected.")
    return True


# This function is separate so we can still print out the results from Puzzle 1
# without having to completely overhaul the return values from the validate_entry_count()
#
# There's definitely more elegant ways to do this, but let's not over-engineer this script -
# it isn't really production code, after all!
def validate_entry_values(entry: dict):
    required_keys = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
    for key in required_keys:
        if key == "byr":
            byr = int(entry[key])
            if byr < 1920 or byr > 2002:
                print(f"Birth year of {byr} is not between 1920 and 2002")
                return False
        elif key == "iyr":
            iyr = int(entry[key])
            if iyr < 2010 or iyr > 2020:
                print(f"Issue year of {iyr} is not between 2010 and 2020")
                return False
        elif key == "eyr":
            eyr = int(entry[key])
            if eyr < 2020 or eyr > 2030:
                print(f"Expiration year of {eyr} is not between 2020 and 2030")
                return False
        elif key == "hgt":
            value = ""
            unit = ""
            stop_checking_numbers = False
            for i in range(len(entry[key])):
                # Probably over-engineered considering the input file
                char = entry[key][i]
                if not stop_checking_numbers and char.isnumeric():
                    value += char
                    continue
                elif not stop_checking_numbers:
                    stop_checking_numbers = True
                unit += char
            value = int(value)

            if unit == "in":
                if value < 59 or value > 76:
                    print("The airline discriminates against people who are too tall or too short,")
                    print(f"and a height of {value} inches is not between 59 and 76")
                    return False
            elif unit == "cm":
                if value < 150 or value > 193:
                    print("The airline discriminates against people who are too tall or too short,")
                    print(f"and a height of {value} cm is not between 150 and 193")
                    return False
            else:
                print(f"Unit {unit} is not either in or cm")
                return False
        elif key == "hcl":
            hcl_valid = True
            if entry[key][0] != "#" or len(entry[key]) != 7:
                hcl_valid = False
            else:
                try:
                    # If this value cannot be converted to hex, i.e. does not consist of 0-9 and a-f,
                    # then this throws a ValueError, telling us this isn't valid
                    int(entry[key][1:], 16)
                except ValueError:
                    hcl_valid = False
            if not hcl_valid:
                print(f"Hair color {entry[key]} does not comply to expected format #nnnnnn where n is a hex number")
                return False
        elif key == "ecl":
            allowed_values = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
            if entry[key] not in allowed_values:
                print(f"Eye color {entry[key]} not in allowed list of colors: {allowed_values}")
                return False
        elif key == "pid":
            if len(entry[key]) != 9:
                print(f"The Passport ID {entry[key]} does not consist of 9 digits.")
                return False
            elif not entry[key].isnumeric():
                print(f"The Passport ID {entry[key]} is not a numeric value.")
                return False
    # If you've made it here, congratulations! It's a valid result.
    print("All entries are valid.")
    return True


def main():
    entries = parse_input("input.txt")
    valid_entries = 0
    valid_entries_2 = 0
    total_entries = 0
    for entry in entries:
        if validate_entry_count(entry):
            valid_entries += 1
            # This way we don't even try to validate the values if some of them are missing,
            # since that means there's already a problem with it.
            if validate_entry_values(entry):
                valid_entries_2 += 1
            print()
        total_entries += 1

    print(f"{valid_entries} out of {total_entries} are valid (Puzzle 1).")
    print(f"{valid_entries_2} out of {total_entries} are valid (Puzzle 2).")


if __name__ == '__main__':
    main()

File: Day 4/Day 4.cpp
#include <vector>
#include <string>
#include <iostream>
#include <fstream>
#include <sstream>

#include "stringsplit.h"

// used to allow us to package both part 1 and part 2 in one function while still being easy to read
enum class valid {
	valid,
	partial,
	invalid
};

class Passport {
public:
	Passport(std::string& raw_fields) {
		raw_fields = strip(raw_fields);
		auto fields = split(raw_fields, " ");
		for (auto& field : fields) {
			field = strip(field);
			auto temp = split(field, ":");
			auto& label = temp[0];
			auto& value = temp[1];
			// Unfortunately I don't think we can do a switch case on a string, so if/if else/if else... it is!
			
			// birth year, issue year, expiration year
			if (label == "byr")
				byr = std::stoi(value);
			else if (label == "iyr")
				iyr = std::stoi(value);
			else if (label == "eyr")
				eyr = std::stoi(value);

			// height - must separate unit from number!
			else if (label == "hgt") {
				std::string number = "";
				std::string unit = "";
				bool adding_to_num = true;
				for (const auto& c : value) {
					if (adding_to_num && std::isdigit(c)) {
						number += c;
						continue;
					}
					else {
						adding_to_num = false;
						unit += c;
					}
				}
				if (number.size() > 0) hgt = std::stoi(number);
				hgt_unit = unit;
			}

			// Hair color, eye color, passport id
			else if (label == "hcl")
				hcl = value;
			else if (label == "ecl")
				ecl = value;
			else if (label == "pid")
				pid = value;
			else if (label == "cid");  // don't care, just don't warn about it
			else std::cout << "Warning: unexpected label found.\n";
		}
	}

	valid validate(std::ofstream* file = nullptr) const {
		if (!check_for_missing_entries()) {
			std::cout << "\tRejected: missing entries.\n\n";
			return valid::invalid;
		}
		else if (!check_for_invalid_entries()) {
			std::cout << "\tRejected: invalid entries.\n\n";
			return valid::partial;
		}
		std::cout << "Accepted.\n\n";
		if (file != nullptr) output_to_file(*file);
		return valid::valid;
	}

private:
	int byr = -1;
	int iyr = -1;
	int eyr = -1;
	int hgt = -1;
	std::string hgt_unit = "";
	std::string hcl = "";
	std::string ecl = "";
	// pid can be represented as a number, but the way we're using it it's more of a string
	std::string pid = "";

	// Debugging helper function.
	[[maybe_unused]] void output_to_file(std::ofstream& file) const {
		file << "byr:" << byr << " iyr:" << iyr << " eyr:" << eyr << " ecl:" << ecl << " hgt:" << hgt << hgt_unit
			<< " hcl:" << hcl << " pid:" << pid << "\n\n";
	}

	bool check_for_missing_entries() const {
		if (byr == -1) {
			std::cout << "Rejected due to missing byr entry.\n";
			return false;
		}
		if (iyr == -1) { 
			std::cout << "Rejected due to missing iyr entry.\n";
			return false;
		}
		if (eyr == -1) {
			std::cout << "Rejected due to missing eyr entry.\n";
			return false;
		}
		if (hgt == -1) {
			std::cout << "Rejected due to missing hgt entry.\n";
			return false;
		}
		if (hcl == "") {
			std::cout << "Rejected due to missing hcl entry.\n";
			return false;
		}
		if (ecl == "") {
			std::cout << "Rejected due to missing ecl entry.\n";
			return false;
		}
		if (pid == "") {
			std::cout << "Rejected due to missing pid entry.\n";
			return false;
		}
		return true;
	}

	bool check_for_invalid_entries() const {
		// endpoints included
		auto entry_between = [&](int min, int max, int val) {
			if (val < min || val > max) return false;
			return true;
		};

		// byr, iyr, eyr are straightforward
		if (!entry_between(1920, 2002, byr)) {
			printf("Rejected: byr %d is not between %d and %d\n", byr, 1920, 2002);
			return false;
		}
		if (!entry_between(2010, 2020, iyr)) {
			printf("Rejected: iyr %d is not between %d and %d\n", iyr, 2010, 2020);
			return false;
		}
		if (!entry_between(2020, 2030, eyr)) {
			printf("Rejected: eyr %d is not between %d and %d\n", byr, 2020, 2030);
			return false;
		}
		
		// hgt requires some checks first
		if (pid == "980577052") {
			std::cout << "STOP\n";
		}
		if (hgt_unit == "in") {
			if (!entry_between(59, 76, hgt)) {
				printf("Rejected: hgt %d in is not between %d and %d\n", hgt, 59, 76);
				return false;
			}
		}
		else if (hgt_unit == "cm") {
			if (!entry_between(150, 193, hgt)) {
				printf("Rejected: hgt %d cm is not between %d and %d\n", hgt, 159, 193);
				return false;
			}
		}
		else {
			std::cout << "Rejected: " << hgt_unit << " is not either in or cm.\n";
			return false;
		}
		// hcl
		if (hcl.size() != 7 || hcl[0] != '#') {
			std::cout << "Rejected: hcl " << hcl << " doesn't conform to expected format #nnnnnn where n is any hex digit\n";
			return false;
		}
		for (int i = 1; i < static_cast<int>(hcl.size()); i++) {
			if (!std::isxdigit(hcl[i])) {
				std::cout << "Rejected: hcl " << hcl << " doesn't conform to expected format #nnnnnn where n is any hex digit\n";
				return false;
			}
		}

		// ecl
		if (ecl != "amb" && ecl != "blu" && ecl != "brn" && ecl != "gry" && ecl != "grn" && ecl != "hzl" && ecl != "oth") {
			std::cout << "Rejected: ecl " << ecl << " isn't allowed.\n";
			return false;
		}

		// pid
			if (pid.size() != 9) {
				std::cout << "Rejected: pid " << pid << " doesn't have exactly 9 digits.\n";
				return false;
			}
		for (const auto& c : pid) {
			if (!std::isdigit(c))
			{
				std::cout << "Rejected: pid " << pid << " isn't a number.\n";
				return false;
			}
		}

		// If you made it here, congratulations! It's (close enough to) valid!
		return true;
	}
};

std::vector<Passport> get_passports(std::string filename) {
	// The Passport object itself is responsible for building itself out of the raw list of fields
	// but we do need to seperate them here
	std::ifstream file(filename);
	std::vector<Passport> passports;
	std::stringstream raw_contents;
	raw_contents << file.rdbuf();
	auto raw_entries = split(raw_contents.str(), "\n\n");
	for (auto& entry : raw_entries) {
		// important cleanup so we can split this in the passport constructor
		for (auto& c : entry) {
			if (c == '\n')
				c = ' ';
		}
		passports.emplace_back(Passport(entry));
	}
	return passports;
}

int main() {
	auto passports = get_passports("input.txt");
	int all_entries_present = 0;
	int all_entries_valid = 0;
	std::ofstream valid_entries("output.txt");
	for (const auto& passport : passports) {
		valid status = passport.validate(&valid_entries);
		if (status == valid::valid || status == valid::partial) {
			if (status == valid::valid) all_entries_valid++;
			all_entries_present++;
		}
	}

	std::cout << all_entries_present << " out of " << passports.size() << " passports have all entries present.\n"
		<< all_entries_valid << " out of " << passports.size() << " passports have all entries valid.\n";
}